File: CP/LeetCode/Python/392.py
# 392. Is Subsequence

from __future__ import annotations


class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        # s_len, t_len = len(s), len(t)
        # s_ptr, t_ptr = 0, 0

        # while s_ptr < s_len and t_ptr < t_len:
        #     if s[s_ptr] == t[t_ptr]:
        #         s_ptr += 1
        #     t_ptr += 1

        # return s_ptr == s_len

        if s == "":
            return True

        s_index = 0
        for char in t:
            if char == s[s_index]:
                s_index += 1
                if s_index == len(s):
                    return True

        return False


if __name__ == "__main__":
    s = "abc"
    t = "ahbgdc"

    solution = Solution()
    print(solution.isSubsequence(s, t))  # Output: True

    s = "axc"
    t = "ahbgdc"

    print(solution.isSubsequence(s, t))  # Output: False

File: CP/LeetCode/Java/392.java
// 392. Is Subsequence

class Solution {

    public boolean isSubsequence(final String s, final String t) {
        int indexS = 0;
        int indexT = 0;

        while (indexS < s.length() && indexT < t.length()) {
            char charS = s.charAt(indexS);
            char charT = t.charAt(indexT);

            if (charS == charT) {
                indexS = indexS + 1;
            }
            indexT = indexT + 1;
        }

        return indexS == s.length();
    }

    public static void main(final String[] args) {
        Solution solution = new Solution();

        System.out.println(solution.isSubsequence("abc", "ahbgdc")); // true
        System.out.println(solution.isSubsequence("axc", "ahbgdc")); // false
    }
}
